Sources/AsyncAlgorithms/AsyncWithLatestFrom2Sequence.swift

```diff
@@ -0,0 +1,261 @@
+      }
+    }
+
+    enum BaseDecision {
+      case pass
+      case returnElement(Result<Element, Error>)
+    }
+
+    var base: Base.AsyncIterator
+    let other1: Other1
+    let other2: Other2
+
+    let othersState: ManagedCriticalState<OthersState>
+
+    var other1Task: Task<Void, Never>?
+    var other2Task: Task<Void, Never>?
+
+    var isTerminated: ManagedCriticalState<Bool>
+
+    // for testability purpose
+    var onBaseElement: (@Sendable (Base.Element) -> Void)?
+    var onOther1Element: (@Sendable (Other1.Element?) -> Void)?
+    var onOther2Element: (@Sendable (Other2.Element?) -> Void)?
+
+    public init(base: Base.AsyncIterator, other1: Other1, other2: Other2) {
+      self.base = base
+      self.other1 = other1
+      self.other2 = other2
+      self.othersState = ManagedCriticalState(.notStarted)
+      self.isTerminated = ManagedCriticalState(false)
+    }
+
+    mutating func startOther1() {
+      let other1IsAlreadyStarted = self.othersState.withCriticalRegion { state -> Bool in
+        switch state.other1State {
+          case .notStarted:
+            state.other1State = .idle
+            return false
+          default:
+            return true
+        }
+      }
+
+      if other1IsAlreadyStarted { return }
+
+      self.other1Task = Task { [other1, othersState, onOther1Element] in
+        do {
+          for try await element in other1 {
+            othersState.withCriticalRegion { state in
+              state.other1State = .element(.success(element))
+            }
+            onOther1Element?(element)
+          }
+        } catch {
+          othersState.withCriticalRegion { state in
+            state.other1State = .element(.failure(error))
+          }
+        }
+      }
+    }
+
+    mutating func startOther2() {
+      let other2IsAlreadyStarted = self.othersState.withCriticalRegion { state -> Bool in
+        switch state.other2State {
+          case .notStarted:
+            state.other2State = .idle
+            return false
+          default:
+            return true
+        }
+      }
+
+      if other2IsAlreadyStarted { return }
+
+      self.other2Task = Task { [other2, othersState, onOther2Element] in
+        do {
+          for try await element in other2 {
+            othersState.withCriticalRegion { state in
+              state.other2State = .element(.success(element))
+            }
+            onOther2Element?(element)
+          }
+        } catch {
+          othersState.withCriticalRegion { state in
+            state.other2State = .element(.failure(error))
+          }
+        }
+      }
+    }
+
+    mutating func startOthers() {
+      self.startOther1()
+      self.startOther2()
+    }
+
+    public mutating func next() async rethrows -> Element? {
+      let shouldReturnNil = self.isTerminated.withCriticalRegion { $0 }
+      guard !shouldReturnNil else { return nil }
+
+      self.startOthers()
```

Comment:
I think we need to start the `Task` in the `makeAsyncIterator` method to kick-off the subscription on the other sequences as fast as possible. This allows the first element from the base to have a higher chance to already have an element from the other sequence